[VSC-7] Make scopes cache optional

heitor-lassarote opened the pull request:
Problem: We were asked during review to make scope cache optional.

Solution: This is accompanied by some changes in the compiler as well.
Turn `scopeCache: ...` argument into `scopeCache?: ...`.

heitor-lassarote committed:
[VSC-7] Make scopes cache optional

Problem: We were asked during review to make scope cache optional.

Solution: This is accompanied by some changes in the compiler as well.
Turn `scopeCache: ...` argument into `scopeCache?: ...`.

GoPavel commented:
LGTM
Could you check suggestions in the https://github.com/dfinity/node-motoko/pull/113 and add some tests on backward compatibility?

heitor-lassarote committed:
[VSC-7] Improve backwards compatibility

Problem: We need to improve backwards compability: if no scope is
passed, the return type should have no scope.

Solution: Create more overloads for `parseMotokoTyped` and write a test
checking that it's working as intended.

GoPavel commented on src/index.ts:
```diff
@@ -63,32 +69,42 @@ export default function wrapMotoko(compiler: Compiler) {
+    ): [ParseMotokoTypedResult | ParseMotokoTypedResult[], ScopeCache]
+        | ParseMotokoTypedResult
+        | ParseMotokoTypedResult[] {
+        if (arguments.length === 1) {
+            if (typeof paths === 'string') {
+                const [progs, _outCache] =
+                    mo.parseMotokoTyped([paths], new Map<string, Scope>());
+                return progs[0];
+            } else {
+                const [progs, _outCache] =
+                    mo.parseMotokoTyped(paths, new Map<string, Scope>());
+                return progs;
+            }
+        }
         if (typeof paths === 'string') {
             const [progs, outCache] = mo.parseMotokoTyped([paths], scopeCache);
```
```suggestion
        if (arguments.length === 1) {
            scopeCache = new Map<string, Scope>();
        }
        if (typeof paths === 'string') {
            const [progs, outCache] = mo.parseMotokoTyped([paths], scopeCache);
```

GoPavel replied:
Why we would do it simpler ?

heitor-lassarote committed:
[VSC-7] Explain backwards compatible code

Problem: It's a little confusing why we do the arguments.length check,
we should explain it.

Solution: Add a comment in both overloads explaining this design.
